Service Manager: Fix order of init/deinit

garbear opened the pull request:
This PR takes #12431 and finishes the task of ordering the individual services in a LIFO manner. Several circular dependencies arose, so some code was split up and refactored.

FernetMenta committed:
game: fix order of init/deinit in ServiceManager

garbear committed:
Revert "Fix crash on startup"

This reverts commit a7d2a4f8169d554dca790a41d86ba53764921d6c.

garbear committed:
Service Manager: Fix order of init/deinit

FernetMenta commented on xbmc/ServiceManager.h:
```diff
@@ -153,6 +157,7 @@ class CServiceManager
   std::unique_ptr<CPlatform> m_Platform;
   std::unique_ptr<PLAYLIST::CPlayListPlayer> m_playlistPlayer;
   std::unique_ptr<CSettings> m_settings;
+  std::unique_ptr<KODI::GAME::CControllerManager> m_controllerManager;
```
how about m_gameControllerManager and GetGameControllerManager? At first glance I was confused about this additional manager because I did not notice that this manager is in the namespace GAME.

garbear replied:
I actually moved this from `CGameServices` because input (and peripherals) depend on some controller stuff, but game services are brought up last. There still exists `CGameServices::GetController*()`, but this is only called from within GAME.

garbear replied:
Oh, I see what you're saying. Just rename it. I could do that, but it would be redundant - Anything outside of game should reference this as `GAME::CControllerManager`

FernetMenta replied:
> Anything outside of game should reference this as GAME::CControllerManager

right but ServiceManager is outside of namaspace game. CServiceManager::GetControllerManager vs CServiceManager::GetGameControllerManager

assume some other namespace has a ControllerManager too. I don't know id C++ allows having the same method more than once with just different return types. 
even if it is allowed, it does not look nice

garbear replied:
> right but ServiceManager is outside of namaspace game. 

Good point. Renamed.

FernetMenta commented:
thanks for cleaning up the mess. feel free to squash my commit into yours.

garbear committed:
Remove g_playlistPlayer pseudonym

Because it made it harder to reason about ServiceManager's dependencies.

garbear committed:
AppParamParser: Fix crash on startup

This fixes a crash on startup introduced in commit 26349e1, "Move
CInputManager into Service Manager" from PR 12366.

garbear commented:
OK, I've tested on win, linux and OSX, and blindly adapted the remaining platforms. Jenkins is building an old version, so when that's failed I'll start another run.

FernetMenta commented:
> Jenkins is building an old version

I noticed this too. Makes our lives a little bit harder :)

MartijnKaijser commented:
if you continuously push new revisions you build up a huge queue so you shouldn't blame jenkins ;)

garbear commented:
I've included a better fix for #12426 that involves wrapping up the remote params in `CAppParamParser` and then unwrapping them inside `CInputManager`, removing the need for calling into `CInputManager` directly.

>> Jenkins is building an old version
> I noticed this too. Makes our lives a little bit harder :)

I'm making liberal use of the `No-Jenkins` flag :)

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
Jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
The `No-Jenkins` tag has tainted the pr. Now jenkins won't touch this with a 10-foot pole.

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
Windows errors look unrelated.

FernetMenta commented on xbmc/ServiceManager.cpp:
```diff
@@ -138,38 +141,57 @@ bool CServiceManager::StartAudioEngine()
 {
+  m_peripherals.reset(new PERIPHERALS::CPeripherals(*m_announcementManager));
   m_peripherals->Initialise();
-  m_PVRManager->Init();
```
there was a reason why this was in stage three. stage 3 is init after WindowManager is initialized. PVR can have its own startup window, hence need to be in stage 3. on the other hand WindowManager depends on peripherials so peripherials must not be in stage 3

garbear replied:
peripherals must not be in stage 2, because our test suite inits stage 2 and doesn't provide enough for the peripherals subsystem

FernetMenta replied:
I care more about the app itself and less about the test suite. I created stage 3 for the reason given.
